File: web/src/features/ApplicationDetail/__tests__/spies/spy-saved-application-repository.ts
import type { SavedApplicationRepository } from '../../../../domain/ports/saved-application-repository';
import type { ApplicationUid, SavedApplication } from '../../../../domain/types';

export class SpySavedApplicationRepository implements SavedApplicationRepository {
  listSavedCalls = 0;
  listSavedResult: readonly SavedApplication[] = [];
  listSavedError: Error | null = null;

  saveCalls: ApplicationUid[] = [];
  saveError: Error | null = null;

  removeCalls: ApplicationUid[] = [];
  removeError: Error | null = null;

  async listSaved(): Promise<readonly SavedApplication[]> {
    this.listSavedCalls += 1;
    if (this.listSavedError) {
      throw this.listSavedError;
    }
    return this.listSavedResult;
  }

  async save(uid: ApplicationUid): Promise<void> {
    this.saveCalls.push(uid);
    if (this.saveError) {
      throw this.saveError;
    }
  }

  async remove(uid: ApplicationUid): Promise<void> {
    this.removeCalls.push(uid);
    if (this.removeError) {
      throw this.removeError;
    }
  }
}

File: mobile/ios/town-crier-tests/Sources/Spies/SpySavedApplicationRepository.swift
import TownCrierDomain

final class SpySavedApplicationRepository: SavedApplicationRepository, @unchecked Sendable {
  private(set) var saveCalls: [PlanningApplication] = []
  var saveResult: Result<Void, Error> = .success(())

  func save(application: PlanningApplication) async throws {
    saveCalls.append(application)
    try saveResult.get()
  }

  private(set) var removeCalls: [String] = []
  var removeResult: Result<Void, Error> = .success(())

  func remove(applicationUid: String) async throws {
    removeCalls.append(applicationUid)
    try removeResult.get()
  }

  private(set) var loadAllCallCount = 0
  var loadAllResult: Result<[SavedApplication], Error> = .success([])

  func loadAll() async throws -> [SavedApplication] {
    loadAllCallCount += 1
    return try loadAllResult.get()
  }
}
